File: exercicios/aula13/ex06_expressoes_minic/expressoes.c
/* Exercício 06 — AST de expressões MINIC respeitando precedência.
 *
 * CLI equivalente a expressoes.py: lê expressões separadas por ';' e imprime a
 * AST de cada uma em notação prefixa. O parser está em expr.c.
 *
 * Uso:  ./expressoes [--pos] <arquivo | ->
 */

#include <stdio.h>
#include <string.h>

#include "expr.h"

int main(int argc, char **argv) {
    int pos = 0, n_args = 0;
    const char *arquivo = NULL;
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--pos") == 0) pos = 1;
        else arquivo = argv[i], n_args++;
    }
    if (n_args != 1) {
        fprintf(stderr, "uso: %s [--pos] <arquivo | ->\n", argv[0]);
        return ERRO_USO;
    }
    Fluxo fl;
    int codigo = fluxo_abrir(&fl, arquivo);
    if (codigo != OK) return codigo;

    while (codigo == OK && !fluxo_verifica(&fl, TOK_EOF)) {
        ExprErro erro;
        Expr *e = expr_parse(&fl, &erro);
        if (!e) {
            erro_sintaxe(erro.token, erro.esperado);
            codigo = ERRO_SINTAXE;
        } else if (!fluxo_aceita(&fl, TOK_SEMICOLON)) {
            erro_sintaxe(fluxo_atual(&fl), "operador ou ';'");
            codigo = ERRO_SINTAXE;
        } else {
            expr_sexpr(stdout, e, pos);
            putchar('\n');
        }
        expr_free(e);
    }
    fluxo_fechar(&fl);
    return codigo;
}

File: exercicios/aula13/ex06_expressoes_minic/expressoes.py
"""Exercício 06 — AST de expressões MINIC respeitando precedência.

Lê uma ou mais expressões separadas por `;` e imprime a AST de cada uma, em
notação prefixa totalmente parentizada. O parser (com as ações que constroem a
árvore) está em expr.py; aqui fica só a CLI.

    a + b * c;     ->  (+ a (* b c))

Com `--pos`, cada folha sai com a posição de origem: `(+ a@1:1 (* b@1:5 c@1:9))`.

Uso:  python expressoes.py [--pos] <arquivo | ->
"""

import sys

import expr as E
from expr import F


def main(argv: list[str]) -> int:
    F.configurar_saida()
    pos = "--pos" in argv[1:]
    args = [a for a in argv[1:] if a != "--pos"]
    if len(args) != 1:
        print("uso: python expressoes.py [--pos] <arquivo | ->", file=sys.stderr)
        return F.ERRO_USO
    fl, codigo = F.abrir(args[0])
    if fl is None:
        return codigo
    parser = E.ExprParser(fl)
    try:
        while not fl.verifica("EOF"):
            arvore = parser.parse_expr()
            fl.exige("SEMICOLON", "operador ou ';'")
            print(E.sexpr(arvore, pos))
    except F.ErroSintaxe as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SINTAXE
    return F.OK


if __name__ == "__main__":
    sys.exit(main(sys.argv))
